pkg/engine/lifecycletest/refresh_test.go

```diff
@@ -1844,3 +1846,119 @@ func TestRefreshWithProgramWithDeletedResource(t *testing.T) {
 }
+
+// Regression test for https://github.com/pulumi/pulumi/issues/19406. Check that if we have a program with
+// more resources than step executor workers we don't deadlock.
+func TestRefreshWithBigProgram(t *testing.T) {
+	t.Parallel()
+
+	programInputs := resource.PropertyMap{"foo": resource.NewStringProperty("bar")}
+	createOutputs := resource.PropertyMap{"foo": resource.NewStringProperty("bar")}
+	readOutputs := resource.PropertyMap{"foo": resource.NewStringProperty("baz")}
+
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				ReadF: func(_ context.Context, req plugin.ReadRequest) (plugin.ReadResponse, error) {
+					if strings.HasPrefix(req.Name, "resA") {
+						assert.Equal(t, createOutputs, req.Inputs)
+						assert.Equal(t, createOutputs, req.State)
+
+						return plugin.ReadResponse{
+							ReadResult: plugin.ReadResult{
+								ID:      req.ID,
+								Inputs:  req.Inputs,
+								Outputs: readOutputs,
+							},
+							Status: resource.StatusOK,
+						}, nil
+					}
+
+					return plugin.ReadResponse{
+						ReadResult: plugin.ReadResult{
+							ID:      req.ID,
+							Inputs:  resource.PropertyMap{},
+							Outputs: resource.PropertyMap{},
+						},
+						Status: resource.StatusOK,
+					}, nil
+				},
+				CreateF: func(_ context.Context, req plugin.CreateRequest) (plugin.CreateResponse, error) {
+					uuid, err := uuid.NewV4()
+					if err != nil {
+						return plugin.CreateResponse{}, err
+					}
+
+					if strings.HasPrefix(req.Name, "resA") {
+						assert.Equal(t, programInputs, req.Properties)
+
+						return plugin.CreateResponse{
+							ID:         resource.ID(uuid.String()),
+							Properties: createOutputs,
+							Status:     resource.StatusOK,
+						}, nil
+					}
+
+					return plugin.CreateResponse{
+						ID:         resource.ID(uuid.String()),
+						Properties: resource.PropertyMap{},
+						Status:     resource.StatusOK,
+					}, nil
+				},
+			}, nil
+		}),
+	}
+
+	programExecutions := 0
+	programF := deploytest.NewLanguageRuntimeF(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		programExecutions++
+
+		for i := int64(0); i < 5; i++ {
+			resp, err := monitor.RegisterResource("pkgA:m:typA", "resA"+strconv.FormatInt(i, 10), true, deploytest.ResourceOptions{
+				Inputs: programInputs,
+			})
+			assert.NoError(t, err)
+
+			// First time we should see the create outputs, second time the read outputs
+			if programExecutions == 1 {
+				assert.Equal(t, createOutputs, resp.Outputs)
+			} else {
+				assert.Equal(t, readOutputs, resp.Outputs)
+			}
+		}
+
+		return nil
+	})
+	hostF := deploytest.NewPluginHostF(nil, nil, programF, loaders...)
+
+	p := &lt.TestPlan{
+		Options: lt.TestUpdateOptions{
+			T:                t,
+			HostF:            hostF,
+			SkipDisplayTests: true,
+			UpdateOptions: engine.UpdateOptions{
+				Parallel: 4,
```

Comment:
Is the key to making this text work that `Parallel: 4`, where 4 is less than the 5 we pick in the loop above? If so, might be worth making this a variable to make the relationship clear e.g. `parallel := 4` and in the loop, for `i := int64(0); i < parallel + 1; i++` or something?